### PR TITLE
Use MSBuildProjectExtensionsPath for RestoreOutputPath

### PR DESCRIPTION
**Pull request by rohit21agrawal**
This PR replaces https://github.com/NuGet/NuGet.Client/pull/2121

This is part of a solution to Microsoft/msbuild#1603. The MSBuildProjectExtensionsPath is where MSBuild will automatically import project .props and .targets files from. This PR changes the RestoreOutputPath to default to MSBuildProjectExtensionsPath instead of BaseIntermediateOutputPath. By default, both of these properties will default to the obj folder, but this change will make the behavior more correct if one of the properties is overridden.

This PR also sets the EnableBaseIntermediateOutputPathMismatchWarning property when appropriate, which enables an MSBuild warning if MSBuildProjectExtensionsPath doesn't match BaseIntermediateOutputPath (added in Microsoft/msbuild#3059).

Included in this PR is also a commit that replaces new string[0] with Array.Empty<string>(), which was something I noticed when browsing the code investigating this change.

CC: @nkolev92 @dsplaisted

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs:42]: We need to make sure that this works in VS. 
We need to make sure this property is correctly evaluated and returned by the ProjectSystem.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectJsonNuGetProject.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectJsonNuGetProject.cs:44]: Likewise, verify that this is correct used in Manual tests as well.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs]: 👏

**Review by nkolev92 — APPROVED** [review]
LGTM, additional tests pending.

Make sure you follow up with new project system to make sure that they pass the correct thing in the nomination, and manually test the msbuildprojectextensionspath use in legacy (not too many tests cover this, cross-plat no op does, but only be accident). 

Please add proper Apex tests for this change.

**Comment by rohit21agrawal** [comment]
@nkolev92 are you saying add tests before we merge? If so, please provide test idea since I do not have enough context on this change.

also I talked @dsplaisted and he mentioned that all changes required for legacy project system are self contained in NuGet, and for dotnet/project-system , he will send out a PR soon. There isn't any synchronization needed as the behavior won't be any worse than it is currently.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs
Comment (rohit21agrawal) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/VsProjectAdapter.cs:42]: I have verified a project.json project from command line and VS and it works perfectly fine.

**Comment by nkolev92** [comment]
@rohit21agrawal 

Regarding the test:

Create a .NET Core Apex test. 
- Create it from template. 
- Unload
- Edit the csproj with some predetermined values, setting the msbuildprojectextensionspath.
- Load project again.
- Install a package with a props/targets, whose props/targets are critical for build. 
- Make sure the project builds. 

Basically make sure you test the reason why we're doing this fix. 

Which is make sure that the props and targets are imported when changing the nuget assets output path or in this case msbuildextensionsoutputpath. 

To make sure your test is good, you can port the same one to current dev (with base intermediate) and expect it to fail. 

Btw, simply creating a NS20 project brings in a package that has targets/props.

File: test/TestUtilities/Test.Utility/compiler/resources/project2.csproj
Comment (jainaashish) [inline test/TestUtilities/Test.Utility/compiler/resources/project2.csproj:2]: 👍

**Review by jainaashish — APPROVED** [review]

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs
Comment (dsplaisted) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs]: It's not valid to set this property here, as it's already been used to import .props from the Microsoft.Common.props.  MSBuild will now generate an error for this (see https://github.com/Microsoft/msbuild/pull/3059), so I would expect test errors once that change makes it to the MSBuild that you use.

The simplest fix is probably to remove `Sdk="Microsoft.NET.Sdk"` from the `Project` tag], and have a `PropertyGroup` that sets this property before an explicit Sdk.props import.  You'll also need an explicit Sdk.targets import at the end of the project file.

Comment (nkolev92) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs]: @dsplaisted 
This test is for dotnet tool...the project style is DotnetToolReference...not packagereference, so there will be no error. 

This is the non-build scenario that we didn't want to break.
So it'll be nice to do it, but it won't cause any problems.

Comment (dsplaisted) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs]: @nkolev92 Changing the `MSBuildProjectExtensionsPath` after `Microsoft.Common.props` has been evaluated will be an error regardless of whether a project is using NuGet, or what the NuGet reference style is.

here will be a warning which NuGet enables if the `BaseIntermediateOutputPath` doesn't match the `MSBuildProjectExtensionsPath` which wouldn't apply here, but the error will still apply.

Comment (nkolev92) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/MsbuildIntegrationTestFixture.cs]: Ahh, true. 

This will cause a failure, when we start using an updated version of CLI for our tests. 
I've created a tracking issue. 

https://github.com/NuGet/Home/issues/6769

@dsplaisted 
When is that change going into CLI?
I think there'll be a change in the template needed in the DotnetTool implementation /cc @wli3

**Pull request merged**
